src/main/java/io/netty/buffer/api/CompositeBuf.java

```diff
@@ -317,61 +316,63 @@ public ByteIterator iterate(int fromOffset, int length) {
         int startBufferIndex = searchOffsets(fromOffset);
         int off = fromOffset - offsets[startBufferIndex];
         Buf startBuf = bufs[startBufferIndex];
-        ByteIterator startIterator = startBuf.iterate(off, Math.min(startBuf.capacity() - off, length));
-        return new ByteIterator() {
+        ByteCursor startCursor = startBuf.iterate(off, Math.min(startBuf.capacity() - off, length));
+        return new ByteCursor() {
             int index = fromOffset;
             final int end = fromOffset + length;
             int bufferIndex = startBufferIndex;
-            ByteIterator itr = startIterator;
+            ByteCursor cursor = startCursor;
+            long longValue = -1;
+            byte byteValue = -1;
 
             @Override
-            public boolean hasNextLong() {
-                return bytesLeft() >= Long.BYTES;
-            }
-
-            @Override
-            public long nextLong() {
-                if (itr.hasNextLong()) {
-                    long val = itr.nextLong();
-                    index += Long.BYTES;
-                    return val;
-                }
-                if (!hasNextLong()) {
-                    throw new NoSuchElementException();
+            public boolean nextLong() {
+                if (bytesLeft() >= Long.BYTES) {
+                    if (cursor.nextLong()) {
+                        longValue = cursor.getLong();
+                        index += Long.BYTES;
+                    } else {
+                        longValue = nextLongFromBytes(); // Leave index increments to 'nextByte'
+                    }
+                    return true;
```

Comment:
Maybe if `ByteCursor` have `setLong` and `setByte` that overwrite the current values, then we could make it the common case to just directly delegate the `nextLong` call, and avoid caching the values in the composite byte cursor. If we can remove `bytesLeft` from the hot path, then we can make it computed and avoid maintaining the `index` field as well.

Reply:
I would prefer to not introduce `set*` methods

Reply:
Do you object to the functionality of overriding the cursor state, or the method names, or both?

Reply:
override the state... I would prefer to keep the state the cursors own business